src/core/raster/qgssinglebandgrayrenderer.cpp

```diff
@@ -218,3 +218,154 @@ QList<int> QgsSingleBandGrayRenderer::usesBands() const
   }
   return bandList;
 }
+
+void QgsSingleBandGrayRenderer::toSld( QDomDocument &doc, QDomElement &element, const QgsStringMap &props ) const
+{
+  QgsStringMap newProps = props;
+
+  // create base structure
+  QgsRasterRenderer::toSld( doc, element, props );
+
+  // look for RasterSymbolizer tag
+  QDomNodeList elements = element.elementsByTagName( QStringLiteral( "sld:RasterSymbolizer" ) );
+  if ( elements.size() == 0)
+    return;
+
+  // there SHOULD be only one
+  QDomElement rasterSymbolizerElem = elements.at(0).toElement();
+
+  // add Channel Selection tags
+  // Need to insert channelSelection in the correct sequence as in SLD standard e.g.
+  // after opacity or geometry or as first element after sld:RasterSymbolizer
+  QDomElement channelSelectionElem = doc.createElement( QStringLiteral( "sld:ChannelSelection" ) );
+  elements = rasterSymbolizerElem.elementsByTagName( QStringLiteral( "sld:Opacity" ) );
+  if ( elements.size() != 0 )
+  {
+    rasterSymbolizerElem.insertAfter( channelSelectionElem, elements.at(0) );
+  }
+  else
+  {
+    elements = rasterSymbolizerElem.elementsByTagName( QStringLiteral( "sld:Geometry" ) );
+    if ( elements.size() != 0 )
+    {
+      rasterSymbolizerElem.insertAfter( channelSelectionElem, elements.at(0) );
+    }
+    else
+    {
+      rasterSymbolizerElem.insertBefore( channelSelectionElem, rasterSymbolizerElem.firstChild() );
+    }
+  }
+
+  // for gray band
+  QDomElement channelElem = doc.createElement( QStringLiteral( "sld:GrayChannel" ) );
+  channelSelectionElem.appendChild( channelElem );
+
+  // set band
+  QDomElement sourceChannelNameElem = doc.createElement( QStringLiteral( "sld:SourceChannelName" ) );
+  sourceChannelNameElem.appendChild( doc.createTextNode( QString::number( grayBand() ) ) );
+  channelElem.appendChild( sourceChannelNameElem );
+
+  // set ContrastEnhancement
+  if ( contrastEnhancement() )
+  {
+    QDomElement contrastEnhancementElem = doc.createElement( QStringLiteral( "sld:ContrastEnhancement" ) );
+    contrastEnhancement()->toSld( doc, contrastEnhancementElem );
+
+    // do changes to minValue/maxValues depending on stretching algorithm. This is necessary because
+    // geoserver do a first stretch on min/max, then apply colo map rules. In some combination is necessary
+    // to use real min/max values and in othere the actual edited min/max values
+    switch ( contrastEnhancement()->contrastEnhancementAlgorithm() )
+    {
+      case ( QgsContrastEnhancement::StretchAndClipToMinimumMaximum ):
+      case ( QgsContrastEnhancement::ClipToMinimumMaximum ):
+      {
+        // with this renderer export have to be check against real min/max values of the raster
+        QgsRasterBandStats myRasterBandStats = mInput->bandStatistics( grayBand(), QgsRasterBandStats::Min | QgsRasterBandStats::Max );
```

Comment:
this call (non cost) has been necessary to have a correct geoserver representation. In case of contrastEnhancement involving Clip, the Normalize sld tags must have real raster min/max values.

Reply:
I think bandStatistics should be made const, and the things it edits should be made mutable. That's the correct approach here.

Reply:
seems now it's ok... back to const and withou using mutable! was my error?

Reply:
@nyalldawson let me know if you need more